### PR TITLE
Add start and end dates for notices

### PR DESCRIPTION
**Pull request by cameron1729**

### CODE REVIEW

File: db/upgrade.php [hunk 1]
```diff
@@ -45,5 +45,25 @@ function xmldb_local_sitenotice_upgrade($oldversion) {
...
+
+            $table = new xmldb_table('local_sitenotice');
+            $field = new xmldb_field('timestart', XMLDB_TYPE_INTEGER, '10', null, XMLDB_NOTNULL, null, '0');
+            $dbman->add_field($table, $field);
+        }
+
+        if (!$dbman->field_exists('local_sitenotice', 'timeend')) {
+
+            $table = new xmldb_table('local_sitenotice');
+            $field = new xmldb_field('timestart', XMLDB_TYPE_INTEGER, '10', null, XMLDB_NOTNULL, null, '0');
```
Comment (dmitriim) [inline db/upgrade.php]: Have you tested that upgrade? :)

Comment (dmitriim) [inline db/upgrade.php:60]: Have you tested that upgrade? :)

**Comment by dmitriim** [comment]
Would be awesome if we can have a notice that is active from specific date and then never expires. Of course this could be achieved by setting up end date far in the future.  Just thinking out loud here.

**Comment by dmitriim** [comment]
Also a quick note from UI testing. Do we really need Is perpetual column? I think we can save space and display only time start and time end. If they are not set, means its perpetual. Thoughts?

image

**Comment by cameron1729** [comment]
Thanks Dima. I've made a few changes (for some reason I can't reply directly to your comments):

1. Got rid of the "Is perpetual" column - you're right, it adds no value to the user
2. Added docblocks where needed
3. Updated unit tests
4. Fixed that upgrade step
5. Having a notice with a start date but no end date is a cool idea (and the other way around too), I'm wondering how we can implement this in the UI? Moodle forms make it a bit challenging. I think it might be a bit out of scope for this issue, but if I have any thoughts about how it could work I'll implement them, if not, I'll create an issue for it

Can you have another look?

**Comment by dmitriim** [comment]
I pulled the latest changes and still can see "Is perpetual" column. Did you forget to push an extra commit?

**Comment by dmitriim** [comment]
When I'm trying to save an exiting notice, I'm getting following error

image

image

File: tests/sitenotice_test.php [hunk 2]
```diff
@@ -78,6 +78,30 @@ private function create_cohort_notice2() {
...
+        $formdata->content = "Date range Notice 1 <a href=\"www.example3.com\">Link 3</a> <a href=\"www.example4.com\">Link 9</a>";
+        $formdata->timestart = time() + HOURSECS;
+        $formdata->timeend = time() + DAYSECS;
+        helper::create_new_notice($formdata);
+    }
+
+    /**
+     * Create sample notice (with a start and end date).
+     */
+    private function create_notice_with_start_and_end2() {
```
Comment (dmitriim) [inline tests/sitenotice_test.php]: I don't think having methods like blabla1 and blabla2 is the best practice. Can you please rewrite to have one generic method where you can pass all required data from your test?

I see that the code already does that. If you can fix existing code that would be a bonus.

**Comment by cameron1729** [comment]
Thanks Dima, I've reworked the unit tests to use data providers. It's not perfect but it's better than it was.

I think the problem about that unexpected property is gone. I tried the following:

1. Make a notice before the patch
2. Upgrade
3. Edit the notice, adding a start and end date

And I didn't see any error.

**Pull request merged**